design_doc.md

```diff
@@ -0,0 +1,268 @@
+
+The API may look like:
+
+```python
+# return a DataArray of DGGS cell centroids as shapely.POINT objects
+ds.dggs.cell_centroids()
+
+# return two DataArrays of DGGS cell centroids as lat/lon coordinates
+ds.dggs.cell_centroids_coords()
+
+# return a DataArray of DGGS cell boundaries as shapely.POLYGON objects
+ds.dggs.cell_boundaries()
+
+# return a DataArray of DGGS cell envelopes as shapely.POLYGON objects
+ds.dggs.cell_envelopes()
+```
+
+## Indexing and Selecting DGGS Data
+
+### Selection by Cell IDs
+
+The simplest way to select DGGS data is by cell ids. This can be done directly using Xarray's API (`.sel()`):
+
+```python
+ds.sel(cell=value)
+```
+
+where `value` can be a single cell id (integer or string/token?) or an array-like of cell ids. This is easily supported by the DGGSIndex encapsulating a PandasIndex. We might also want to support other `value` types, e.g.,
+
+- assuming that DGGS cell ids are defined such that contiguous cells in space have contiguous id values, we could provide a `slice` to define a range of cell ids.
+- DGGSIndex might implement some DGGS-aware logic such that it auto-detects if the given input cells are parent cells (lower DGGS resolution) and then selects all child cells accordingly.
+
+We might want to select cell neighbors (i.e., return a new Dataset/DataArray with a new neighbor dimension), probably via a specific API (`.dggs` accessors).
+
+### Selection by Geometries (Spatial Indexing)
+
+Another useful way of selecting DGGS data is from input geometries (spatial queries), e.g.,
+
+- Select all cells that are the closest to a collection of data points
+- Select all cells that intersects with or are fully contained in a polygon
+
+This kind of selection requires spatial indexes as this can not be done with a pandas index (see [this issue](https://github.com/benbovy/xdggs/issues/16)).
+
+If we support spatial indexing directly in `xdggs`, we can hardly reuse Xarray's `.sel()` for spatial queries as `ds.sel(cell=shapely.Polygon(...))` would look quite confusing. Perhaps better would be to align with [xvec](https://github.com/xarray-contrib/xvec) and have a separate `.dggs.query()` method.
+
+Alternatively, we could just get away with the conversion and cell geometry extraction methods proposed above and leave spatial indexes/queries to [xvec](https://github.com/xarray-contrib/xvec).
+
+## Operations between similar DGGS (alignment)
+
+TODO
+
+## Plotting
+
+Three approaches are possible (non-mutually exclusive):
+
+1. convert cell data into gridded or raster data (choose grid/raster resolution depending on the resolution of the rendered figure) and then reuse existing python plotting libraries (matplotlib, cartopy) maybe through xarray plotting API
+2. convert cell data into vector data and plot the latter via, e.g., [xvec](https://github.com/xarray-contrib/xvec) or [geopandas](https://github.com/geopandas/geopandas) API
```

Comment:
Do you mean hvplot plotting vector data like https://hvplot.holoviz.org/reference/geopandas/polygons.html?

Reply:
I meant this: https://discourse.holoviz.org/t/dealing-high-resolution-vector-plots-xarray-holoviz/6310

Reply:
Thanks for the links! Yes it looks like a similar approach to what I suggest by using Datashader with rasterization of DGGS to a dynamic resolution and extent. This could actually be applied for polygon plots: downgrade the DGGS resolution (#18), then aggregate data, then convert it to vector data, then plot it. It would be nice to have some high-level API to do that interactively instead of requiring the user doing all of those steps manually.

Reply:
I rephrased the paragraph below in [5a97e2c](https://github.com/benbovy/xdggs/pull/20/commits/5a97e2c7c405bc8bc9cfdfd801b1cd330046a5ab) to reflect that comment.